<a href="https://colab.research.google.com/github/AlienX77-cmd/Python-for-Business-and-Finance/blob/main/Part_4_Exercise_06.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Coding Exercises (Part 6)

Now, you will have the opportunity to analyze your own dataset. <br>
__Follow the instructions__ and insert your code! You are either requested to
- Complete the Code and __Fill in the gaps__. Gaps are marked with "__---__" and are __placeholders__ for your code fragment.
- Write Code completely __on your own__

In some exercises, you will find questions that can only be answered, if your code is correct and returns the right output! The correct answer is provided below your coding cell. There you can check whether your code is correct.

If you need a hint, check the __Hints Section__ at the end of this Notebook. Exercises and Hints are numerated accordingly.

If you need some further help or if you want to check your code, you can also watch the __solutions videos__ or check the __solutions notebook__.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

## Pandas: Intermediate

### Exercise 6: NA Values & Summary Statistics

In [1]:
# run the cell!
import pandas as pd
pd.options.display.float_format = '{:.2f}'.format

In [3]:
# run the cell!
listings = pd.read_csv("listings.csv")

In [4]:
# run the cell!
listings.head()

Symbol Exchange                           Name  Last_Price     Market_Cap  \
0      A     NYSE     Agilent Technologies, Inc.       81.68 25934696179.92   
1     AA     NYSE              Alcoa Corporation       29.15  5407809589.60   
2   AABA   NASDAQ                    Altaba Inc.       75.39 42781131315.30   
3    AAC     NYSE             AAC Holdings, Inc.        2.16    53141086.80   
4    AAL   NASDAQ  American Airlines Group, Inc.       34.02 15276869742.96   

   ADR TSO  IPO_Year            Sector  \
0      NaN   1999.00     Capital Goods   
1      NaN   2016.00  Basic Industries   
2      NaN       NaN        Technology   
3      NaN   2014.00       Health Care   
4      NaN       NaN    Transportation   

                                           Industry  
0  Biotechnology: Laboratory Analytical Instruments  
1                                          Aluminum  
2                                      EDP Services  
3                              Medical Specialities  
4                     Air Freight/Delivery Services

In [5]:
# run the cell
listings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6852 entries, 0 to 6851
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Symbol      6852 non-null   object 
 1   Exchange    6852 non-null   object 
 2   Name        6852 non-null   object 
 3   Last_Price  6745 non-null   float64
 4   Market_Cap  5954 non-null   float64
 5   ADR TSO     140 non-null    float64
 6   IPO_Year    3105 non-null   float64
 7   Sector      5309 non-null   object 
 8   Industry    5309 non-null   object 
dtypes: float64(4), object(5)
memory usage: 481.9+ KB


51. Get the __amount__ of __NA-values__ per __column__! __Fill in the gaps__! <br> How many __NA-values__ are in the __ADR TSO__ column?

In [7]:
listings.isna().sum(axis = 0)

Symbol           0
Exchange         0
Name             0
Last_Price     107
Market_Cap     898
ADR TSO       6712
IPO_Year      3747
Sector        1543
Industry      1543
dtype: int64

There are ... 6712 NA-values in the ADR TSO column!

52. Get the __amount__ of __non-null values__ per column (don´t use info() method)! <br>

In [9]:
listings.notna().sum(axis = 0)

Symbol        6852
Exchange      6852
Name          6852
Last_Price    6745
Market_Cap    5954
ADR TSO        140
IPO_Year      3105
Sector        5309
Industry      5309
dtype: int64

53. __Drop__ all __rows__ with __at least one NA-Value__ from listings (don´t overwrite listings!). __How many__ rows/listing are __left__?

In [11]:
listings.dropna().shape

(100, 9)

There are only ... 100 rows left.

54. __Drop__ all __columns__ with __at least one NA-Value__ from listings (don´t overwrite listings!). __How many__ columns are __left__?

In [12]:
listings.dropna(axis = 1).shape

(6852, 3)

There are only ... 3 columns left!

55. We finally decide to __drop__ all __columns__ with __less than 3000 non-null values__. __Overwrite__ listings! __Fill in the gaps!__

In [13]:
listings.dropna(axis = 1, thresh = 3000, inplace = True) #drop columns with more than 3000 null values

In [14]:
# run the cell!
listings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6852 entries, 0 to 6851
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Symbol      6852 non-null   object 
 1   Exchange    6852 non-null   object 
 2   Name        6852 non-null   object 
 3   Last_Price  6745 non-null   float64
 4   Market_Cap  5954 non-null   float64
 5   IPO_Year    3105 non-null   float64
 6   Sector      5309 non-null   object 
 7   Industry    5309 non-null   object 
dtypes: float64(3), object(5)
memory usage: 428.4+ KB


56. __Filter__ listings for all rows/listings with __NA-values__ in the __Last_Price__ column. Save the new DataFrame in the variable __no_price__. <br>
__Fill in the gaps__!

In [16]:
no_price = listings.loc[listings.Last_Price.isna()]

What do all listings with __NA-values in the Last_Price__ column have in common (hint: check the __Market_Cap__ column)?

In [17]:
# run the cell and inspect!
no_price.head()

Symbol Exchange                                       Name  Last_Price  \
151   AGM.A     NYSE  Federal Agricultural Mortgage Corporation         NaN   
190  AIG.WS     NYSE         American International Group, Inc.         NaN   
224   AKO.A     NYSE                  Embotelladora Andina S.A.         NaN   
225   AKO.B     NYSE                  Embotelladora Andina S.A.         NaN   
254    ALIT   NASDAQ                                Alight Inc.         NaN   

     Market_Cap  IPO_Year Sector Industry  
151        0.00       NaN    NaN      NaN  
190        0.00       NaN    NaN      NaN  
224        0.00       NaN    NaN      NaN  
225        0.00       NaN    NaN      NaN  
254        0.00       NaN    NaN      NaN

In [18]:
# run the cell and inspect!
no_price.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 107 entries, 151 to 6839
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Symbol      107 non-null    object 
 1   Exchange    107 non-null    object 
 2   Name        107 non-null    object 
 3   Last_Price  0 non-null      float64
 4   Market_Cap  107 non-null    float64
 5   IPO_Year    27 non-null     float64
 6   Sector      3 non-null      object 
 7   Industry    3 non-null      object 
dtypes: float64(3), object(5)
memory usage: 7.5+ KB


In [19]:
#run the cell and inspect!
no_price.Market_Cap.value_counts()

0.00    107
Name: Market_Cap, dtype: int64

The Market_Cap is 0.0 for all listings with NA-values in the Last_Price column. Therefore, it might make sense to...

57. ...__Fill all NA-Values__ in the __Last_Price__ column of listings with __0.0__! __Save the change__!

In [20]:
listings.Last_Price.fillna(0, inplace = True)

In [21]:
# run the cell!
listings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6852 entries, 0 to 6851
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Symbol      6852 non-null   object 
 1   Exchange    6852 non-null   object 
 2   Name        6852 non-null   object 
 3   Last_Price  6852 non-null   float64
 4   Market_Cap  5954 non-null   float64
 5   IPO_Year    3105 non-null   float64
 6   Sector      5309 non-null   object 
 7   Industry    5309 non-null   object 
dtypes: float64(3), object(5)
memory usage: 428.4+ KB


58. __Filter__ listings for all rows/listings that have __non-null values__ in the __IPO_Year__ column! __Save__ the new DataFrame in the variable __known_year__. <br>
__Fill in the gaps!__

In [23]:
known_year = listings.loc[listings.IPO_Year.notna()]

__How many__ listings with known IPO_Year are in the __known_year__ DataFrame?

In [24]:
# run the cell!
known_year.head()

Symbol Exchange                           Name  Last_Price      Market_Cap  \
0       A     NYSE     Agilent Technologies, Inc.       81.68  25934696179.92   
1      AA     NYSE              Alcoa Corporation       29.15   5407809589.60   
3     AAC     NYSE             AAC Holdings, Inc.        2.16     53141086.80   
8    AAOI   NASDAQ  Applied Optoelectronics, Inc.       12.30    244556139.90   
11   AAPL   NASDAQ                     Apple Inc.      200.62 945979473600.00   

    IPO_Year            Sector  \
0    1999.00     Capital Goods   
1    2016.00  Basic Industries   
3    2014.00       Health Care   
8    2013.00        Technology   
11   1980.00        Technology   

                                            Industry  
0   Biotechnology: Laboratory Analytical Instruments  
1                                           Aluminum  
3                               Medical Specialities  
8                                     Semiconductors  
11                            Computer Manufacturing

In [25]:
# run the cell!
known_year.shape

(3105, 8)

There are ... 3105 listings with known IPO_Year!

Finally, we figured out that __IPO_Year__ information is __not available__ for stocks with an IPO __prior to 1970__. Therefore...

59. ... we decide to __fill all NA-Values__ in the __IPO_Year__ column with __1969__. __Save__ the change!

In [26]:
listings.IPO_Year.fillna(1969, inplace = True)

Make a __final inspection__ of our listings DataFrame. Which __columns__ still contain __NA-values__?

In [27]:
# run the cell!
listings.head()

Symbol Exchange                           Name  Last_Price     Market_Cap  \
0      A     NYSE     Agilent Technologies, Inc.       81.68 25934696179.92   
1     AA     NYSE              Alcoa Corporation       29.15  5407809589.60   
2   AABA   NASDAQ                    Altaba Inc.       75.39 42781131315.30   
3    AAC     NYSE             AAC Holdings, Inc.        2.16    53141086.80   
4    AAL   NASDAQ  American Airlines Group, Inc.       34.02 15276869742.96   

   IPO_Year            Sector  \
0   1999.00     Capital Goods   
1   2016.00  Basic Industries   
2   1969.00        Technology   
3   2014.00       Health Care   
4   1969.00    Transportation   

                                           Industry  
0  Biotechnology: Laboratory Analytical Instruments  
1                                          Aluminum  
2                                      EDP Services  
3                              Medical Specialities  
4                     Air Freight/Delivery Services

In [28]:
#run the cell!
listings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6852 entries, 0 to 6851
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Symbol      6852 non-null   object 
 1   Exchange    6852 non-null   object 
 2   Name        6852 non-null   object 
 3   Last_Price  6852 non-null   float64
 4   Market_Cap  5954 non-null   float64
 5   IPO_Year    6852 non-null   float64
 6   Sector      5309 non-null   object 
 7   Industry    5309 non-null   object 
dtypes: float64(3), object(5)
memory usage: 428.4+ KB


The columns ... Sector & Industry still contain NA-values!

60. __Transform the datatype__ in the IPO_Year column to integer (beyond the scope of this course, run the cell!)

In [29]:
#run the cell!
listings.IPO_Year = listings.IPO_Year.astype("int")

In [30]:
#run the cell and inspect!
listings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6852 entries, 0 to 6851
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Symbol      6852 non-null   object 
 1   Exchange    6852 non-null   object 
 2   Name        6852 non-null   object 
 3   Last_Price  6852 non-null   float64
 4   Market_Cap  5954 non-null   float64
 5   IPO_Year    6852 non-null   int64  
 6   Sector      5309 non-null   object 
 7   Industry    5309 non-null   object 
dtypes: float64(2), int64(1), object(5)
memory usage: 428.4+ KB


In [31]:
#run the cell!
listings.head()

Symbol Exchange                           Name  Last_Price     Market_Cap  \
0      A     NYSE     Agilent Technologies, Inc.       81.68 25934696179.92   
1     AA     NYSE              Alcoa Corporation       29.15  5407809589.60   
2   AABA   NASDAQ                    Altaba Inc.       75.39 42781131315.30   
3    AAC     NYSE             AAC Holdings, Inc.        2.16    53141086.80   
4    AAL   NASDAQ  American Airlines Group, Inc.       34.02 15276869742.96   

   IPO_Year            Sector  \
0      1999     Capital Goods   
1      2016  Basic Industries   
2      1969        Technology   
3      2014       Health Care   
4      1969    Transportation   

                                           Industry  
0  Biotechnology: Laboratory Analytical Instruments  
1                                          Aluminum  
2                                      EDP Services  
3                              Medical Specialities  
4                     Air Freight/Delivery Services

61. __Save and export__ the cleaned DataFrame to the new csv-file __listings_clean.csv__! __Fill in the gaps!__

In [32]:
listings.to_csv("listings_clean.csv", index = False)

62. __Re-Import__ listings_clean.csv and save the DataFrame in the variable __listings_clean__!

In [34]:
listings_clean = pd.read_csv("listings_clean.csv")

In [35]:
# run the cell!
listings_clean.head()

Symbol Exchange                           Name  Last_Price     Market_Cap  \
0      A     NYSE     Agilent Technologies, Inc.       81.68 25934696179.92   
1     AA     NYSE              Alcoa Corporation       29.15  5407809589.60   
2   AABA   NASDAQ                    Altaba Inc.       75.39 42781131315.30   
3    AAC     NYSE             AAC Holdings, Inc.        2.16    53141086.80   
4    AAL   NASDAQ  American Airlines Group, Inc.       34.02 15276869742.96   

   IPO_Year            Sector  \
0      1999     Capital Goods   
1      2016  Basic Industries   
2      1969        Technology   
3      2014       Health Care   
4      1969    Transportation   

                                           Industry  
0  Biotechnology: Laboratory Analytical Instruments  
1                                          Aluminum  
2                                      EDP Services  
3                              Medical Specialities  
4                     Air Freight/Delivery Services

In [36]:
# run the cell!
listings_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6852 entries, 0 to 6851
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Symbol      6852 non-null   object 
 1   Exchange    6852 non-null   object 
 2   Name        6852 non-null   object 
 3   Last_Price  6852 non-null   float64
 4   Market_Cap  5954 non-null   float64
 5   IPO_Year    6852 non-null   int64  
 6   Sector      5309 non-null   object 
 7   Industry    5309 non-null   object 
dtypes: float64(2), int64(1), object(5)
memory usage: 428.4+ KB


63. Call an appropriate method on the __listings_clean__ DataFrame that provides plenty of __summary statistics__!

In [38]:
listings_clean.describe()

Last_Price      Market_Cap  IPO_Year
count     6852.00         5954.00   6852.00
mean        36.79   7174440219.09   1987.14
std         98.58  35254552925.59     20.89
min          0.00            0.00   1969.00
25%          7.70    120376889.75   1969.00
50%         19.98    558517593.27   1969.00
75%         36.85   2888746939.46   2012.00
max       4299.99 945979473600.00   2019.00

63a. __Calculate__ the __mean__ and the __standard deviation__ for the numerical columns (use the appropriate method!).

In [39]:
listings_clean.agg(["mean", "std"])

<ipython-input-39-5ed26ebfc507>:1: FutureWarning: ['Symbol', 'Exchange', 'Name', 'Sector', 'Industry'] did not aggregate successfully. If any error is raised this will raise in a future version of pandas. Drop these columns/ops to avoid this warning.
  listings_clean.agg(["mean", "std"])


Last_Price     Market_Cap  IPO_Year
mean       36.79  7174440219.09   1987.14
std        98.58 35254552925.59     20.89

64. Get some summary statistics on all __non-numerical columns__ by passing __"O"__ to the __include parameter__ of the __describe()__ method! __Fill in the gaps!__ <br>
What is the __most frequent__ value in the __Industry__ column?

In [40]:
listings_clean.describe(include = "O")

Symbol Exchange            Name   Sector               Industry
count    6852     6852            6852     5309                   5309
unique   6840        3            5766       12                    135
top      CDOR   NASDAQ  Public Storage  Finance  Major Pharmaceuticals
freq        2     3435              14     1059                    470

The most frequent Industry is ... Major Pharmaceuticals with 470 appearances.

65. Get the __maximum__ values for all columns! Which is the (alphabetically) __last Symbol__?

In [41]:
listings_clean.max()

<ipython-input-41-acd42b5cf42a>:1: FutureWarning: The default value of numeric_only in DataFrame.max is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  listings_clean.max()


Symbol                  ZYXI
Exchange                NYSE
Name             voxeljet AG
Last_Price           4299.99
Market_Cap   945979473600.00
IPO_Year                2019
dtype: object

The (alphabetically) last Symbol is ... ZYXI!

66. __Calculate__ the __total Market_Cap__ over all listings.

In [42]:
listings_clean.Market_Cap.sum()

42716617064473.484

67. Create a __correlation matrix__ for the __listings_clean__ DataFrame! Are Market_Cap and Last_Price __positively correlated__?

In [43]:
listings_clean.corr()

<ipython-input-43-257041b11356>:1: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  listings_clean.corr()


Last_Price  Market_Cap  IPO_Year
Last_Price        1.00        0.31     -0.06
Market_Cap        0.31        1.00     -0.07
IPO_Year         -0.06       -0.07      1.00

Market_Cap and Last_Price are ... positively correlated.The higher the Price, the higher the Market Capitalization.

# Well Done!

------------------------------

# Hints (Spoiler!)

51. isna() method, sum() method

52. notna() method, sum() method

53. dropna(axis = 0) method and shape attribute

54. dropna(axis = 1) method and shape attribute

55. dropna() method, thresh parameter

56. pass listings.Last_Price.isna() to the loc operator

57. fillna() method

58. pass listings.IPO_Year.notna() to the loc operator

59. fillna() method

60. run the cell!

61. to_csv() method

62. You should know this already!

63. describe() method

63a. agg() method, pass list with "mean" and "std"

64. describe() method and include parameter

65. max() method

66. sum() method

67. corr() method